client/my-sites/theme/controller.jsx

```diff
@@ -9,33 +9,48 @@ import ThemeNotFoundError from './theme-not-found-error';
 const debug = debugFactory( 'calypso:themes' );
 
 export function fetchThemeDetailsData( context, next ) {
-	if ( ! context.isServerSide || context.cachedMarkup ) {
+	if ( context.cachedMarkup ) {
 		return next();
 	}
 
 	const themeSlug = context.params.slug;
 	const theme = getTheme( context.store.getState(), 'wpcom', themeSlug );
+	const themeDotOrg = getTheme( context.store.getState(), 'wporg', themeSlug );
 
-	if ( theme ) {
-		debug( 'found theme!', theme.id );
+	if ( theme || themeDotOrg ) {
+		debug( 'found theme!', theme?.id ?? themeDotOrg.id );
 		return next();
 	}
 
 	context.store
 		.dispatch( requestTheme( themeSlug, 'wpcom', context.lang ) )
 		.then( () => {
 			const themeDetails = getTheme( context.store.getState(), 'wpcom', themeSlug );
-			if ( ! themeDetails ) {
-				const error = getThemeRequestErrors( context.store.getState(), themeSlug, 'wpcom' );
-				debug( `Error fetching theme ${ themeSlug } details: `, error.message || error );
-				const err = {
-					status: 404,
-					message: 'Theme Not Found',
-					themeSlug,
-				};
-				return next( err );
+			if ( themeDetails ) {
+				return next();
 			}
-			next();
+
+			context.store
+				.dispatch( requestTheme( themeSlug, 'wporg', context.lang ) )
+				.then( () => {
+					const themeOrgDetails = getTheme( context.store.getState(), 'wporg', themeSlug );
+					if ( ! themeOrgDetails ) {
+						const err = {
+							status: 404,
+							message: 'Theme Not Found',
+							themeSlug,
+						};
+						const error = getThemeRequestErrors( context.store.getState(), themeSlug, 'wporg' );
+						debug( `Error fetching WPORG theme ${ themeSlug } details: `, error.message || error );
+						return next( err );
+					}
+
+					next();
+				} )
+				.catch( next );
+
+			const error = getThemeRequestErrors( context.store.getState(), themeSlug, 'wpcom' );
+			debug( `Error fetching WPCOM theme ${ themeSlug } details: `, error.message || error );
```

Comment:
I would recommend dispatching this in parallel to improve performance. API requests can be slow! Something similar to this:

```js
// "all" instead of "race" so that we avoid going to .then if an error is the first thing to happen.
Promise.all( [
  context.store.dispatch( requestTheme... ),
  context.store.dispatch( requestTheme... )
]).then( () => {
 // ... here we can check for both wporg or wpcom themes and handle error conditions.
} )
```

Reply:
Actually, that's how I did it the first time but decided against it because that will make a request to .ORG APIs for all themes, even if the theme is managed by WPCOM. 

Since these endpoints are available to logged-out visitors, I thought we should make an API request .org only if it's necessary. :)

Reply:
That's a fair point :)

It'd be really nice if the API handled this for us, but that'd probably be a lot of effort for not much benefit! 😬